src/app/CommandResponseSender.h

```diff
@@ -51,91 +93,79 @@ class CommandResponseSender : public Messaging::ExchangeDelegate
      *         exchange context has been assigned or the context
      *         has been released.
      */
-    Messaging::ExchangeContext * GetExchangeContext() const { return mExchangeCtx.Get(); }
-
-    /**
-     * @brief Flush acks right now, typically done when processing a slow command
-     */
-    void FlushAcksRightNow()
-    {
-        VerifyOrReturn(mExchangeCtx);
-        auto * msgContext = mExchangeCtx->GetReliableMessageContext();
-        VerifyOrReturn(msgContext != nullptr);
-        msgContext->FlushAcks();
-    }
+    Messaging::ExchangeContext * GetExchangeContext() const override { return mExchangeCtx.Get(); }
 
     /**
      * Gets subject descriptor of the exchange.
      *
      * WARNING: This method should only be called when the caller is certain the
      * session has not been evicted.
      */
-    Access::SubjectDescriptor GetSubjectDescriptor() const
+    Access::SubjectDescriptor GetSubjectDescriptor() const override
     {
         VerifyOrDie(mExchangeCtx);
         return mExchangeCtx->GetSessionHandle()->GetSubjectDescriptor();
     }
 
-    bool HasSessionHandle() { return mExchangeCtx && mExchangeCtx->HasSessionHandle(); }
-
-    FabricIndex GetAccessingFabricIndex() const
+    FabricIndex GetAccessingFabricIndex() const override
     {
         VerifyOrDie(mExchangeCtx);
         return mExchangeCtx->GetSessionHandle()->GetFabricIndex();
     }
 
-    void SetExchangeContext(Messaging::ExchangeContext * ec) { mExchangeCtx.Grab(ec); }
-
-    void WillSendMessage() { mExchangeCtx->WillSendMessage(); }
-
-    bool IsForGroup()
-    {
-        VerifyOrDie(mExchangeCtx);
-        return mExchangeCtx->IsGroupExchangeContext();
-    }
-
-    bool HasExchangeContext() { return mExchangeCtx.Get() != nullptr; }
-
-    GroupId GetGroupId()
+    Optional<GroupId> GetGroupId() const override
     {
         VerifyOrDie(mExchangeCtx);
-        return mExchangeCtx->GetSessionHandle()->AsIncomingGroupSession()->GetGroupId();
+        if (!mExchangeCtx->IsGroupExchangeContext())
+        {
+            return NullOptional;
+        }
+        return MakeOptional(mExchangeCtx->GetSessionHandle()->AsIncomingGroupSession()->GetGroupId());
```

Comment:
Worth a comment that explains why we are guaranteed this is an _incoming_ group session.

Alternately, we should test for the session type, not the context type, if that's what we care about.

Reply:
I mean `mExchangeCtx->IsGroupExchangeContext()` is just doing `return mSession && mSession->IsGroupSession()`. For readability sake I will just check `sessionHandle->IsGroupSession()` here.